tracing-opentelemetry/src/layer.rs

```diff
@@ -0,0 +1,276 @@
+///
+/// [tracing]: https://github.com/tokio-rs/tracing
+pub struct OpenTelemetryLayer<S, T: api::Tracer> {
+    tracer: T,
+
+    get_context: WithContext,
+    _registry: marker::PhantomData<S>,
+}
+
+// this function "remembers" the types of the subscriber so that we
+// can downcast to something aware of them without knowing those
+// types at the callsite.
+//
+// See https://github.com/tokio-rs/tracing/blob/4dad420ee1d4607bad79270c1520673fa6266a3d/tracing-error/src/layer.rs
+pub(crate) struct WithContext(
+    fn(&tracing::Dispatch, &span::Id, f: &mut dyn FnMut(&mut api::SpanBuilder)),
+);
+
+impl WithContext {
+    // This function allows a function to be called in the context of the
+    // "remembered" subscriber.
+    pub(crate) fn with_context<'a>(
+        &self,
+        dispatch: &'a tracing::Dispatch,
+        id: &span::Id,
+        mut f: impl FnMut(&mut api::SpanBuilder),
+    ) {
+        (self.0)(dispatch, id, &mut f)
+    }
+}
+
+pub(crate) fn build_context(builder: &mut api::SpanBuilder) -> api::SpanContext {
+    let span_id = builder.span_id.expect("Builders must have id");
+    let (trace_id, trace_flags) = builder
+        .parent_context
+        .as_ref()
+        .map(|parent_context| (parent_context.trace_id(), parent_context.trace_flags()))
+        .unwrap_or_else(|| {
+            (
+                builder.trace_id.expect("trace_id should exist"),
+                api::TRACE_FLAG_SAMPLED,
+            )
+        });
+
+    api::SpanContext::new(trace_id, span_id, trace_flags, false)
+}
+
+struct SpanEventVisitor<'a>(&'a mut api::Event);
+
+impl<'a> field::Visit for SpanEventVisitor<'a> {
+    /// Record events on the underlying OpenTelemetry `Span`.
+    fn record_debug(&mut self, field: &field::Field, value: &dyn fmt::Debug) {
+        if field.name() == "message" {
+            self.0.name = format!("{:?}", value);
+        } else {
+            self.0
+                .attributes
+                .push(api::Key::new(field.name()).string(format!("{:?}", value)));
+        }
+    }
+}
+
+struct SpanAttributeVisitor<'a>(&'a mut api::SpanBuilder);
+
+impl<'a> field::Visit for SpanAttributeVisitor<'a> {
+    /// Set attributes on the underlying OpenTelemetry `Span`.
+    fn record_debug(&mut self, field: &field::Field, value: &dyn fmt::Debug) {
+        let attribute = api::Key::new(field.name()).string(format!("{:?}", value));
+        if let Some(attributes) = &mut self.0.attributes {
+            attributes.push(attribute);
+        } else {
+            self.0.attributes = Some(vec![attribute]);
+        }
+    }
+}
+
+impl<S, T> OpenTelemetryLayer<S, T>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    T: api::Tracer + 'static,
+{
+    /// Retrieve the parent OpenTelemetry [`SpanContext`] from the current
+    /// tracing [`span`] through the [`Registry`]. This [`SpanContext`]
+    /// links spans to their parent for proper hierarchical visualization.
+    fn parent_context(
```

Comment:
I believe that these links would need references if they're actually rendered. Since this is a private API, it doesn't matter.